### PR TITLE
Basic Routing

### PR DESCRIPTION
**Pull request by aaronmgdr**
1 Selecting Parks from the List or map now changes the location hash
2 Changing the location hash sets the app state park
3 the tree takes us home without a reload

### CODE REVIEW

**Comment by aaronmgdr** [comment]
ive deployed this branch to gh-pages

**Comment by aaronmgdr** [comment]
#17

**Comment by johntyree** [comment]
Nice.

It doesn't appear to remember filter settings, but it's still way better than refreshing the entire page :+1:.

File: client/js/components/App.react.jsx [hunk 1]
```diff
@@ -128,14 +128,15 @@ export default class App extends React.Component {
...
-                    selectPark={(park) => this.selectPark(park)}
-                    selectParkWithId={(parkId) => this.selectParkWithId(parkId)}
                     applyFilters={(filter) => this.applyFilters(filter)} />
             );
         }
 
         return (
-            <div>{content}</div>
+            <div>
+                <AppRouter selectParkWithId={this.selectParkWithId.bind(this)}/>
```
Comment (luqmaan) [inline client/js/components/App.react.jsx]: You can also do

```
(parkId) => this.selectParkWithId(parkId)
```

Instead of binding

File: client/js/components/AppRouter.jsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+
+export default class AppRouter extends React.Component {
+
+  componentWillMount() {
```
Comment (luqmaan) [inline client/js/components/AppRouter.jsx]: Inconsistent indentation. Use 4 spaces. Also everywhere else in the rest of this file, pls

File: client/js/components/AllParksMap.jsx [hunk 3]
```diff
@@ -47,7 +47,7 @@ export default class AllParksMap extends React.Component {
         });
 
         layer.on('click', () => {
-            this.props.onSelectPark(feature.id);
+           window.location.hash = feature.id
```
Comment (luqmaan) [inline client/js/components/AllParksMap.jsx]: Inconsistent indentation.

**Comment by luqmaan** [comment]
Currently the route for a park is

```
/#394
```

It should be

```
/park/#394
```

File: client/js/components/AppRouter.jsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+   componentDidUnmount() {
+       window.removeEventListener('hashchange', this.onHashChange.bind(this));
+   }
+
+   onHashChange() {
+      this.props.selectParkWithId(window.location.hash.substring(1));
+   }
+
+   render () {
```
Comment (luqmaan) [inline client/js/components/AppRouter.jsx]: Should be

`render() {`

Not

`render () {`

File: client/js/components/AppRouter.jsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+     return null;
```
Comment (luqmaan) [inline client/js/components/AppRouter.jsx]: Can just do `return;`

**Comment by luqmaan** [comment]
Works great. :deciduous_tree: after those style tweaks

**Comment by prestonp** [comment]
Hey good job on the routing! 

I wanted to point out that the `AppRouter` detects clicks on the map or the park list, but not url hashes when the page is first loaded. It should route pages to parks like `http://localhost:1234/#167` since folks might copy and share links. 

I think the two best courses of action are to:

##### 1. Make AppRouter the top component and handle window.location.hash upon mounting

Here's a [snippet](https://github.com/rackt/react-router/blob/master/doc/00%20Guides/0%20Overview.md#without-react-router) showing how to handle routing manually. 

Basically, the state for the router is just the window.location.hash and the router.render determines which subcomponent is plugged in. Unfortunately, it kinda sucks if you have nested resources because you have to do extra work to parse ids like 

`/users/34/favorites/4`

##### 2. Use a plug-in like react-router

This might be a little overkill since theres only two pages atm, but it has the benefit of using regular url paths like `http://localhost:1234/parks/167` instead of using hashes. It also has a fallback for browsers that don't support the html5 history api.

**Comment by luqmaan** [comment]
We have to use hashes because of github pages.

**Comment by prestonp** [comment]
Ah @luqmaan you're right, nvm den

**Comment by luqmaan** [comment]
Thanks pham.

Been waiting so long to make that pun.

**Comment by luqmaan** [comment]
:deciduous_tree: you should merge before this pr becomes stale!

> AppRouter detects clicks on the map or the park list, but not url hashes when the page is first loaded

@prestonp or @fremn do you want to open a new issue with this?

**Comment by prestonp** [comment]
Yeah, it's probably best to do another issue, I could take a stab at it after this branch gets merged

**Comment by aaronmgdr** [comment]
ill fix those style things and merge in today.

**Comment by aaronmgdr** [comment]
@luqmaan i think setting the initial state from the hash and then using SetState on hash change should fix it right?

**Comment by aaronmgdr** [comment]
in the interest of merging id say new issue though

**Comment by luqmaan** [comment]
:deciduous_tree:
